### PR TITLE
AArch64: Enable Register Association

### PR DESCRIPTION
**Pull request by Akira1Saitoh**
Implement Register Association for aarch64.

- Introduce `assocreg` pseudo instruction.
- Keep track of the real register to virtual register mappings
  which are set up by register dependencies.
- Emit `assocreg` instructions at the end of every extended basic block.
- Emit `assocreg` instructions when setting up register dependencies.
- In register assignment phase, update virtual register to real register mappings
  when we encounter an `assocreg` instruction.

- Enhance `findBestFreeRegister` and `freeBestRegister` to use register assocations.

Signed-off-by: Akira Saitoh <saiaki@jp.ibm.com>

### CODE REVIEW

File: compiler/aarch64/codegen/OMRCodeGenerator.cpp [hunk 1]
```diff
@@ -625,3 +625,9 @@ OMR::ARM64::CodeGenerator::generateScratchRegisterManager(int32_t capacity)
    {
    return new (self()->trHeapMemory()) TR_ARM64ScratchRegisterManager(capacity, self());
    }
+
+bool
+OMR::ARM64::CodeGenerator::enableRegisterAssociations()
+   {
+   return true;
```
Comment (knn-k) [inline compiler/aarch64/codegen/OMRCodeGenerator.cpp:632]: Can you move this function body to OMRCodeGenerator.hpp if it only returns `true`.

File: compiler/aarch64/codegen/OMRMachine.hpp [hunk 2]
```diff
@@ -62,14 +62,30 @@ class OMR_EXTENSIBLE Machine : public OMR::Machine
...
     * @brief Finds the best free register
     * @param[in] rk : register kind
     * @param[in] considerUnlatched : consider unlatched state or not
     * @return Free RealRegister
     */
    TR::RealRegister *findBestFreeRegister(TR_RegisterKinds rk,
-                                            bool considerUnlatched = false);
+                                            bool considerUnlatched = false)
+      {
+      return findBestFreeRegister(NULL, rk, false);
```
Comment (knn-k) [inline compiler/aarch64/codegen/OMRMachine.hpp:87]: Is it OK to pass `false` as the third argument here?
I guess it should be `considerUnlatched` instead.

Comment (Akira1Saitoh) [reply compiler/aarch64/codegen/OMRMachine.hpp]: Thanks for pointing it out. Fixed.

File: compiler/aarch64/codegen/OMRCodeGenerator.cpp (hunk 1 above)
Comment (Akira1Saitoh) [reply compiler/aarch64/codegen/OMRCodeGenerator.cpp]: Moved it to OMRCodeGenerator.hpp.

File: compiler/aarch64/codegen/ARM64Debug.cpp [hunk 3]
```diff
@@ -1489,6 +1496,32 @@ TR_Debug::print(TR::FILE *pOutFile, TR::RegisterDependencyConditions *conditions
...
+void
+TR_Debug::printAssocRegDirective(TR::FILE *pOutFile, TR::Instruction *instr)
+   {
+   TR_ARM64RegisterDependencyGroup * depGroup = instr->getDependencyConditions()->getPostConditions();
+
+   printPrefix(pOutFile, instr);
+   trfprintf(pOutFile, "%s", getOpCodeName(&instr->getOpCode()));
+   trfflush(pOutFile);
+
+   int first = TR::RealRegister::FirstGPR;
```
Comment (knn-k) [inline compiler/aarch64/codegen/ARM64Debug.cpp:1508]: This variable `first` is not used in this function.

Comment (Akira1Saitoh) [reply compiler/aarch64/codegen/ARM64Debug.cpp]: Removed `first` and `last` and rewrote the loop to use `getAddCursorForPost()` for the loop end condition.

File: compiler/aarch64/codegen/OMRMachine.cpp [hunk 4]
```diff
@@ -33,15 +33,87 @@
...
+      if ((conditions = cursor->getDependencyConditions()) != NULL)
+         {
+         TR::Register * boundReg = conditions->searchPostConditionRegister(realReg);
+         if (boundReg == NULL)
+            {
+            boundReg = conditions->searchPreConditionRegister(realReg);
+            }
+         if (boundReg != NULL)
+            {
+            if (boundReg == virtReg)
```
Comment (knn-k) [inline compiler/aarch64/codegen/OMRMachine.cpp:79]: This part can be simply `return (boundReg == virtReg);`.

File: compiler/aarch64/codegen/OMRMachine.cpp [hunk 5]
```diff
@@ -97,13 +207,31 @@ TR::RealRegister *OMR::ARM64::Machine::freeBestRegister(TR::Instruction *current
...
+
    if (forced != NULL)
       {
       best = forced;
       candidates[0] = best->getAssignedRegister();
       }
    else
       {
+      uint32_t interference = 0; // TODO implement live register analysis for aarch64
+      uint32_t preference = 0, pref_favored = 0;
```
Comment (knn-k) [inline compiler/aarch64/codegen/OMRMachine.cpp:220]: I think it is better to declare `pref_favored` as a variable of type `bool`.

File: compiler/aarch64/codegen/OMRMachine.cpp (hunk 4 above)
Comment (Akira1Saitoh) [reply compiler/aarch64/codegen/OMRMachine.cpp]: Thanks. I changed it as you suggested.

File: compiler/aarch64/codegen/OMRMachine.cpp (hunk 5 above)
Comment (Akira1Saitoh) [reply compiler/aarch64/codegen/OMRMachine.cpp]: Changed `pref_favored` type to `bool`.

**Comment by knn-k** [comment]
@genie-omr build aarch64

**Comment by knn-k** [comment]
There are some failures in x86-64 macOS, but they don't seem to have been caused by this PR:
```
2020-09-03T13:21:52.5760670Z 31: [  PASSED  ] 233 tests.
2020-09-03T13:21:52.5761030Z 31: [  FAILED  ] 2 tests, listed below:
2020-09-03T13:21:52.5761440Z 31: [  FAILED  ] PortSockTest.poll_functionality_basic
2020-09-03T13:21:52.5761900Z 31: [  FAILED  ] PortSockTest.poll_functionality_many_sockets
```

**Comment by knn-k** [comment]
@genie-omr build aarch64

**Review by knn-k — APPROVED** [review]

**Pull request merged**